<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Описание проекта</a></span></li></ul></li><li><span><a href="#Создание-подключения-к-БД" data-toc-modified-id="Создание-подключения-к-БД-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Создание подключения к БД</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Решение-задач" data-toc-modified-id="Решение-задач-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Решение задач</a></span><ul class="toc-item"><li><span><a href="#Посчитать,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитать,-сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитать, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитать количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определить издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Определить-автора-с-самой-высокой-средней-оценкой-книг-(для-книг-с-50-и-более-оценками)" data-toc-modified-id="Определить-автора-с-самой-высокой-средней-оценкой-книг-(для-книг-с-50-и-более-оценками)-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определить автора с самой высокой средней оценкой книг (для книг с 50 и более оценками)</a></span></li><li><span><a href="#Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Проект по SQL - Анализ базы данных книг

### Описание проекта
Компания купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать его базу данных. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель исследования:**
1. *Определить направление развития книжного сервиса.*

**Описание данных**

В базе данных содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.

Таблица `books` (данные о книгах):

    book_id — идентификатор книги;
    author_id — идентификатор автора;
    title — название книги;
    num_pages — количество страниц;
    publication_date — дата публикации книги;
    publisher_id — идентификатор издателя.
    
Таблица `authors` (данные об авторах):

    author_id — идентификатор автора;
    author — имя автора.
    
Таблица `publishers` (данные об издательствах):

    publisher_id — идентификатор издательства;
    publisher — название издательства.
    
Таблица `ratings` (данные о пользовательских оценках книг):

    rating_id — идентификатор оценки;
    book_id — идентификатор книги;
    username — имя пользователя, оставившего оценку;
    rating — оценка книги.
    
Таблица `reviews` (данные о пользовательских обзорах):

    review_id — идентификатор обзора;
    book_id — идентификатор книги;
    username — имя автора обзора;
    text — текст обзора.

**Схема данных**

![data schema](https://pictures.s3.yandex.net/resources/scheme_1589269096.png)

**Задачи:**

* Посчитать, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитать количество обзоров и среднюю оценку;
* Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
* Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
* Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Создание подключения к БД
Создадим подключение к базе данных:

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры
db_config = {'user': 'login', # имя пользователя
'pwd': 'password', # пароль
'host': 'localhost',
'port': 6432, # порт подключения
'db': 'books-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

Создадим функцию `select` для выполнения SQL-запросов:

In [2]:
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

## Обзор данных
Исследуем таблицы и выведем их первые строки:

In [3]:
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']
for table in tables:
    print(f'\nТаблица: {table}')
    sql = '''SELECT *
             FROM {}
             LIMIT 5'''.format(table)
    display(select(sql))


Таблица: books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


Таблица: authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


Таблица: publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


Таблица: ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


Таблица: reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Мы можем наблюдать, что:
* данные представлены как количественными (`num_pages`, `rating`), так и категориальными (`author`, `publisher`) величинами;
* все данные стандартизированы и в предобработке не нуждаются;
* в основной таблице `books` столбцы `author` и `publisher` представлены в виде числовых кодировок, расшифровка которых содержится в соответствующих таблицах;
* в таблицах `ratings` и `reviews` не содержится наименований книг - только их идентификаторы `book_id`, отсылающие нас к основной таблице `books`.

## Решение задач

### Посчитать, сколько книг вышло после 1 января 2000 года
Запишем в `sql` соответствующий запрос и выведем результат на экран:

In [4]:
sql = '''SELECT COUNT(book_id) AS books_total
         FROM books
         WHERE publication_date > '2000-01-01';
         '''
select(sql)

books_total
0          819

В нашей БД содержатся данные о 819 книгах, вышедших после 1 января 2000 года.

### Для каждой книги посчитать количество обзоров и среднюю оценку
Запишем в `sql` соответствующий запрос и выведем результат на экран:

In [5]:
sql = '''SELECT 
             b.title AS title,
             COUNT(DISTINCT re.review_id) AS reviews_total,
             AVG(ra.rating) AS avg_rating
         FROM books AS b
         INNER JOIN reviews AS re ON re.book_id = b.book_id
         INNER JOIN ratings AS ra ON ra.book_id = b.book_id
         GROUP BY 
              b.title, 
              b.book_id
        ORDER BY reviews_total DESC;'''

select(sql)

title  reviews_total  avg_rating
0                Twilight (Twilight  #1)              7    3.662500
1                               The Road              6    3.772727
2                         The Book Thief              6    4.264151
3                       The Glass Castle              6    4.206897
4                    Water for Elephants              6    3.977273
..                                   ...            ...         ...
989    Naked Empire (Sword of Truth  #8)              1    3.500000
990                  Moo  Baa  La La La!              1    3.000000
991  Merrick (The Vampire Chronicles #7)              1    4.000000
992                            Babyville              1    3.500000
993                           Lysistrata              1    4.000000

[994 rows x 3 columns]

Мы можем наблюдать, что в БД содержится около 1000 книг, причём количество отзывов в них далеко не всегда коррелируется с их рейтингом. 

В качестве примера можно рассмотреть вампирскую тематику. В ней есть как книга с большим количеством обзоров, но с не самой высокой оценкой (`Twilight (Twilight #1)`), так и книга, где всего один обзор, но оценена она при этом выше (`Merrick (The Vampire Chronicles #7)`).

### Определить издательство, которое выпустило наибольшее число книг толще 50 страниц
Запишем в `sql` соответствующий запрос и выведем результат на экран:

In [6]:
sql = '''SELECT 
             p.publisher as publisher, 
             COUNT(b.book_id) AS books_total
         FROM books AS b
         INNER JOIN publishers AS p ON b.publisher_id = p.publisher_id
         WHERE b.num_pages > 50
         GROUP BY publisher
         ORDER BY books_total DESC
         LIMIT 1;'''

select(sql)

publisher  books_total
0  Penguin Books           42

Лидером является одно из старейших британских издательств - Penguin Books. Специализируется оно в основном на романах и серьёзной литературе - в своё время там были изданы «Прощай, оружие» Э. Хемингуэя, «Любовник леди Чаттерлей» Д. Г. Лоуренса и др.

### Определить автора с самой высокой средней оценкой книг (для книг с 50 и более оценками)
Запишем в `sql` соответствующий запрос и выведем результат на экран:

In [7]:
sql = '''WITH many_ratings AS
  (SELECT b.book_id,
          b.author_id,
          a.author,
          AVG(ra.rating) AS rating_avg
   FROM books AS b
   INNER JOIN ratings AS ra ON b.book_id = ra.book_id
   INNER JOIN authors AS a ON b.author_id = a.author_id
   GROUP BY b.book_id, a.author_id
   HAVING COUNT(ra.rating_id) >= 50)
   
SELECT author, 
       AVG(rating_avg) as rating
FROM many_ratings
GROUP BY author
ORDER BY rating DESC
LIMIT 1;
'''

select(sql)

author    rating
0  J.K. Rowling/Mary GrandPré  4.283844

Согласно нашей БД, наиболее признанным читателями являются книги Дж. К. Роулинг с иллюстрациями Мэри Грандпре. Именно эти издания саги о Гарри Поттере являются самыми известными, в том числе и у нас в стране.

### Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок
Запишем в `sql` соответствующий запрос и выведем результат на экран:

In [8]:
sql = '''WITH review_count AS
           (SELECT COUNT(re.text) AS reviews_count
            FROM reviews AS re
            FULL OUTER JOIN
              (SELECT username,
                      COUNT(rating_id) AS rating_count
               FROM ratings
               GROUP BY username) AS ra_c ON ra_c.username = re.username
            WHERE rating_count > 50
            GROUP BY re.username)
   
SELECT ROUND(AVG(reviews_count)) AS avg_reviews_count
FROM review_count;'''

select(sql)

avg_reviews_count
0               24.0

Среднее количество обзоров от таких активных пользователей равно 24.

## Общий вывод

Мы провели краткий анализ данных сервиса для чтения книг по подписке. 

Исходя из информации, полученной из БД:

* Всего в ней содержится 819 современных книг (т.е. выпущенных на последние 20 лет).

* Практически на 1000 книг есть обзоры читателей и их оценки, которые, впрочем, далеко не всегда коррелируются между собой. 

  Если планируется внедрение рекомендательной системы, то внимание следует сосредоточить сперва на количестве отзывов, а уже потом - на средней оценке. Нередки случаи, когда у книг присутствует один-два высоких отзыва, и опираться лишь на них будет не вполне объективно.
  
* Лидером по выпуску внушительных фолиантов является одно из старейших британских издательств - Penguin Books. В своё время там были изданы «Прощай, оружие» Э. Хемингуэя, «Любовник леди Чаттерлей» Д. Г. Лоуренса и другие известные романы.

  Если убрать из запроса к базе данных ограничение, то мы можем получить список издательств, книги которых можно рекомендовать любителям серьёзной классической литературы.
  
* В большинстве своём читатели отдают предпочтение более лёгкой литературе - особенно распространены иллюстрированные издания. Бесспорным фаворитом является сага о Гарри Поттере.

  Опять же, слегка видоизменив запрос к базе данных, можно получить список наиболее популярных авторов и их средний читательский рейтинг. Данная информация также может быть использована в рекомендательной системе - к примеру, в качестве варианта по умолчанию для тех пользователей, кто не указал свои предпочтения.
  
  
* Среднее количество обзоров от пользователей, которые поставили больше 50 оценок, - 24. Поскольку именно такие активные пользователи являются бесплатной рекламой не только книг, но и самого сервиса, для них стоит разработать систему поощрений. К примеру, можно:
    * установить для них специальный статус (вида "Читатель-эксперт") с соответствующим значком в профиле и рамкой на аватаре;
    * предложить бесплатный месячный доступ к некоторой выборке предположительно-перспективных книг взамен на 1-2 подробные рецензии.